File: v4/rust/iprange-livedb/src/validation/bitmap/word_cache.rs
use crate::bitmap_page;
use crate::error::{Error, Result};

use super::{coverage, require_query_header, required_level, Kind};
use crate::validation::context::Context;
use crate::validation::ValidationSink;

pub(crate) struct WordCache {
    root: u32,
    limit: u64,
    kind: Kind,
    leaf_base: Option<u64>,
    leaf_page: Option<u32>,
    present: bool,
}

impl WordCache {
    pub(crate) fn new(root: u32, limit: u64, kind: Kind) -> Self {
        Self {
            root,
            limit,
            kind,
            leaf_base: None,
            leaf_page: None,
            present: false,
        }
    }

    pub(crate) fn word<S: ValidationSink>(
        &mut self,
        context: &Context<'_, S>,
        word_index: u32,
    ) -> Result<u64> {
        let bit = u64::from(word_index)
            .checked_mul(64)
            .ok_or(Error::ArithmeticOverflow("validation bitmap word"))?;
        if bit >= self.limit || self.root == 0 {
            return Ok(0);
        }
        let leaf_base = bit / bitmap_page::LEAF_BITS * bitmap_page::LEAF_BITS;
        if self.leaf_base != Some(leaf_base) {
            self.load_leaf(context, bit, leaf_base)?;
        }
        if !self.present {
            return Ok(0);
        }
        let local_word = usize::try_from((bit - leaf_base) / 64)
            .map_err(|_| Error::ArithmeticOverflow("validation bitmap word"))?;
        let page_number = self
            .leaf_page
            .ok_or(Error::Corrupt("validation bitmap leaf is missing"))?;
        let page = context.mapping.page(page_number, context.meta.page_count)?;
        require_query_header(page, context.meta.txn_id, self.kind, 0)?;
        bitmap_page::leaf_word(page, local_word)
    }

    fn load_leaf<S: ValidationSink>(
        &mut self,
        context: &Context<'_, S>,
        bit: u64,
        leaf_base: u64,
    ) -> Result<()> {
        let mut page_number = self.root;
        let mut level = required_level(self.limit)?;
        let mut base = 0u64;
        loop {
            let page = context.mapping.page(page_number, context.meta.page_count)?;
            require_query_header(page, context.meta.txn_id, self.kind, level)?;
            if level == 0 {
                self.leaf_base = Some(base);
                self.leaf_page = Some(page_number);
                self.present = true;
                return Ok(());
            }
            let (child, child_base) = self.selected_child(page, bit, base, level)?;
            if child == 0 {
                self.leaf_base = Some(leaf_base);
                self.leaf_page = None;
                self.present = false;
                return Ok(());
            }
            base = child_base;
            page_number = child;
            level -= 1;
        }
    }

    fn selected_child<P: crate::mapping::ByteSource>(
        &self,
        page: P,
        bit: u64,
        base: u64,
        level: u16,
    ) -> Result<(u32, u64)> {
        let span = coverage(level - 1)?;
        let index = usize::try_from((bit - base) / span)
            .map_err(|_| Error::ArithmeticOverflow("validation bitmap word"))?;
        if index >= bitmap_page::BRANCH_CHILDREN {
            return Err(Error::Corrupt(
                "validated bitmap child is outside its branch",
            ));
        }
        let child = bitmap_page::branch_child(page, index)?;
        let child_base = base
            .checked_add(
                span.checked_mul(index as u64)
                    .ok_or(Error::ArithmeticOverflow("validation bitmap word"))?,
            )
            .ok_or(Error::ArithmeticOverflow("validation bitmap word"))?;
        Ok((child, child_base))
    }
}

File: v4/go/internal/validation/word_cache.go
package validation

// Bitmap word cache of the membership scan (Rust validation/bitmap/
// word_cache.rs): one cached leaf of word reads over the raw mapping,
// outside the graph claims. Every read re-verifies the walked headers
// (Rust require_query_header) and folds any defect into the Corrupt
// class, which the membership scan collapses into its own finding.

import (
	"github.com/firehol/iprange/v4/go/internal/bitmap"
	"github.com/firehol/iprange/v4/go/internal/format"
)

// bitmapWordCache is the one-leaf cache (Rust WordCache).
type bitmapWordCache struct {
	root     uint32
	limit    uint64
	kind     bitmap.Kind
	leafBase uint64
	leafPage uint32
	cached   bool
	present  bool
}

func newBitmapWordCache(root uint32, limit uint64, kind bitmap.Kind) bitmapWordCache {
	return bitmapWordCache{root: root, limit: limit, kind: kind}
}

// word reads one global 64-bit word of the bitmap (Rust WordCache::word:
// out-of-range words and absent roots read empty).
func (w *bitmapWordCache) word(ctx *context, wordIndex uint32) (uint64, error) {
	bit := uint64(wordIndex) * 64
	if bit >= w.limit || w.root == 0 {
		return 0, nil
	}
	leafBase := bit / bitmap.LeafBits * bitmap.LeafBits
	if !w.cached || w.leafBase != leafBase {
		if err := w.loadLeaf(ctx, bit); err != nil {
			return 0, err
		}
	}
	if !w.present {
		return 0, nil
	}
	page, err := ctx.mapping.Page(w.leafPage)
	if err != nil {
		return 0, err
	}
	level := uint16(0)
	if _, err := checkedBitmapHeader(ctx, page, w.kind, &level); err != nil {
		return 0, err
	}
	return bitmap.LeafWord(page, int((bit-leafBase)/64))
}

// loadLeaf resolves and caches the leaf page covering bit (Rust
// WordCache::load_leaf: the descent re-proves every header and an absent
// child answers empty).
func (w *bitmapWordCache) loadLeaf(ctx *context, bit uint64) error {
	level, err := bitmap.RequiredLevel(w.limit)
	if err != nil {
		return formatError(err)
	}
	pageNumber := w.root
	base := uint64(0)
	for {
		page, err := ctx.mapping.Page(pageNumber)
		if err != nil {
			return err
		}
		header, err := checkedBitmapHeader(ctx, page, w.kind, &level)
		if err != nil {
			return err
		}
		if header.Level == 0 {
			w.leafBase = base
			w.leafPage = pageNumber
			w.cached = true
			w.present = true
			return nil
		}
		span, err := bitmap.Coverage(header.Level - 1)
		if err != nil {
			return formatError(err)
		}
		index := (bit - base) / span
		if index >= uint64(bitmap.BranchChildren) {
			return &format.Error{Code: format.CodeFormatInvalid, Detail: "validated bitmap child is outside its branch"}
		}
		child, err := bitmap.BranchChild(page, int(index))
		if err != nil {
			return formatError(err)
		}
		childBase, err := bitmapWordBase(base, span, index)
		if err != nil {
			return err
		}
		if child == 0 {
			w.cached = true
			w.present = false
			return nil
		}
		base = childBase
		pageNumber = child
		level--
	}
}

// bitmapWordBase computes the child base bit of one branch slot (Rust
// WordCache::selected_child base arithmetic).
func bitmapWordBase(base, span uint64, index uint64) (uint64, error) {
	offset := span * index
	if span != 0 && offset/span != index {
		return 0, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation bitmap word"}
	}
	childBase := base + offset
	if childBase < base {
		return 0, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation bitmap word"}
	}
	return childBase, nil
}

// checkedBitmapHeader re-verifies one bitmap page header of a raw
// cross-check read (Rust require_query_header; any change is the Corrupt
// class folded into the caller's finding).
func checkedBitmapHeader(ctx *context, page []byte, kind bitmap.Kind, expectedLevel *uint16) (bitmap.Header, error) {
	header, problem := bitmap.CheckedHeader(page, ctx.meta.TxnID, kind, expectedLevel)
	if problem != bitmap.HeaderProblemNone {
		return bitmap.Header{}, &format.Error{Code: format.CodeFormatInvalid, Detail: "validated bitmap changed during cross-check"}
	}
	return header, nil
}
